src/main/java/com/mongodb/spark/sql/connector/read/partitioner/PartitionerHelper.java

```diff
@@ -101,21 +107,43 @@ public static List<BsonDocument> createPartitionPipeline(
   public static BsonDocument storageStats(final ReadConfig readConfig) {
     LOGGER.info("Getting collection stats for: {}", readConfig.getNamespace().getFullName());
     try {
-      return readConfig
-          .withCollection(
-              coll -> Optional.ofNullable(coll.aggregate(COLL_STATS_AGGREGATION_PIPELINE)
-                      .allowDiskUse(readConfig.getAggregationAllowDiskUse())
-                      .comment(readConfig.getComment())
-                      .first())
-                  .orElseGet(BsonDocument::new))
-          .getDocument("storageStats", new BsonDocument());
-    } catch (RuntimeException ex) {
-      if (ex instanceof MongoCommandException
-          && (ex.getMessage().contains("not found.")
-              || ((MongoCommandException) ex).getCode() == 26)) {
+      return readConfig.withCollection(
+          coll -> Optional.ofNullable(coll.aggregate(COLL_STATS_AGGREGATION_PIPELINE)
+                  .allowDiskUse(readConfig.getAggregationAllowDiskUse())
+                  .comment(readConfig.getComment())
+                  .first())
+              .orElseGet(BsonDocument::new));
+    } catch (MongoCommandException ex) {
+      if (ex.getMessage().contains("not found.") || ex.getCode() == 26) {
         LOGGER.info("Could not find collection: {}", readConfig.getCollectionName());
         return new BsonDocument();
       }
+
+      // Atlas Data Federation does not support the storageStats property and requires
+      // special handling to return the federated collection stats.
+      if (ex.getMessage().contains("Data Federation") || ex.getCode() == 9) {
+        return storageStatsDataFederation(readConfig);
+      }
+      throw new MongoSparkException("Partitioner calling collStats command failed", ex);
+    } catch (RuntimeException ex) {
+      throw new MongoSparkException("Partitioner calling collStats command failed", ex);
+    }
+  }
+
+  private static BsonDocument storageStatsDataFederation(final ReadConfig readConfig) {
+    try {
+      return readConfig.withCollection(
+          coll -> Optional.ofNullable(coll.aggregate(COLL_STATS_DATA_FEDERATION_AGGREGATION_PIPELINE)
+                  .allowDiskUse(readConfig.getAggregationAllowDiskUse())
+                  .comment(readConfig.getComment())
+                  .map(collStats -> collStats.append(
+                      "size",
+                      collStats
+                          .getDocument("partition", new BsonDocument())
+                          .getNumber("size", new BsonInt32(0))))
+                  .first())
```

Comment:
One issue with that approach is that when you have multiple partitions (S3 or Mongo clusters) backing up your federated collection, you'll only use the info from the first partition. Whereas the `collStats` command gives you the total size of all partitions.

I think this can be mitigated by summing the partition size and count of all partitions, then returning a `BsonDocument` from that.

Reply:
Thanks @guillotjulien - according to the [atlas docs](https://www.mongodb.com/docs/atlas/data-federation/supported-unsupported/pipeline/collstats/)

> The following example shows $collStats syntax for retrieving the total number of documents in the partitions.
> ```
> use s3Db
> db.abc.aggregate([ {$collStats: {"count" : {} }} ])
> ```

Have you found that is not the case?

Reply:
@rozza it does indeed, but it gives you one document per partition as shown in the example above. So if you want the total, you'd need to compute the sum of counts of all partitions.

e.g.
```
[
  { $collStats: { "count" : {} } },
  { $group: { _id: null, totalCount: { $sum: "$count" }, totalSize: { $sum: "$partition.size" } }  }
]
```

Reply:
Thanks @guillotjulien I've added that to the latest commit.